src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs

```diff
@@ -0,0 +1,313 @@
+        MessagePackSerializer.WriteStr8Header(buffer, 0, validNameLength);
+
+        return cursor;
+    }
+
+    private byte[] ResolveTableMappingForCategoryName(string categoryName)
+    {
+        var tableNameCache = this.m_tableNameCache;
+
+        if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+        {
+            return tableName;
+        }
+
+        return this.ResolveTableMappingForCategoryNameRare(categoryName);
+    }
+
+    private byte[] ResolveTableMappingForCategoryNameRare(string categoryName)
+    {
+        lock (this.m_lockObject)
+        {
+            var tableNameCache = this.m_tableNameCache;
+
+            // Check if another thread added the mapping while we waited on the
+            // lock.
+            if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+            {
+                return tableName;
+            }
+
+            byte[] mappedTableName = null;
+
+            // If user configured table name mappings run resolution logic.
+            if (this.m_tableMappings != null
+                && !this.m_tableMappings.TryGetValue(categoryName, out mappedTableName))
+            {
+                // Find best match if an exact match was not found.
+
+                string currentKey = null;
+
+                foreach (var mapping in this.m_tableMappings)
+                {
+                    if (!categoryName.StartsWith(mapping.Key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    if (currentKey == null || mapping.Key.Length >= currentKey.Length)
+                    {
+                        currentKey = mapping.Key;
+                        mappedTableName = mapping.Value;
+                    }
+                }
+            }
+
+            mappedTableName ??= !this.m_shouldPassThruTableMappings
+                ? this.m_defaultTableName
+                : s_passthroughTableName;
+
+            if (mappedTableName.Length <= 0 && tableNameCache.CachedSanitizedTableNameCount < MaxCachedSanitizedTableNames)
+            {
+                // We resolved to a wildcard which is pass-through mode.
+
+                Span<byte> stackBuffer = stackalloc byte[64];
+
+                int bytesWritten = SanitizedCategoryName(stackBuffer, categoryName);
+                if (bytesWritten > 0)
+                {
+                    mappedTableName = stackBuffer.Slice(0, bytesWritten).ToArray();
+
+                    tableNameCache.CachedSanitizedTableNameCount++;
+                }
+                else
+                {
+                    // Note: When the table name could not be sanitized we cache
+                    // the empty array NOT s_passthroughTableName.
+                    mappedTableName = Array.Empty<byte>();
+                }
+            }
+
+            // Note: This is using copy-on-write pattern to keep the happy
```

Comment:
Could we use ` ConcurrentDictionary`? The reads are lock-free for a `ConcurrentDictionary`.

Reply:
@utpilla I replied to this above.